src/database.rs

```diff
@@ -499,183 +519,205 @@ impl Database {
-            let scope = self.scope_result().clone();
-            let name = self.name_result().clone();
+    pub fn export(&self, stmt: &str) {
+        time(&self.options, "export", || {
             let sref = self.statement(stmt).unwrap_or_else(|| {
                 panic!("Label {} did not correspond to an existing statement", stmt)
             });
 
             File::create(format!("{}.mmp", stmt))
                 .map_err(export::ExportError::Io)
-                .and_then(|mut file| export::export_mmp(&parse, &name, &scope, sref, &mut file))
+                .and_then(|mut file| self.export_mmp(sref, &mut file))
                 .unwrap()
         })
     }
 
     /// Export the grammar of this database in DOT format.
     #[cfg(feature = "dot")]
-    pub fn export_grammar_dot(&mut self) {
-        time(&self.options.clone(), "export_grammar_dot", || {
-            let name = self.name_result().clone();
-            let grammar = self.grammar_result().clone();
+    pub fn export_grammar_dot(&self) {
+        time(&self.options, "export_grammar_dot", || {
+            let name = self.name_result();
+            let grammar = self.grammar_result();
 
             File::create("grammar.dot")
                 .map_err(export::ExportError::Io)
-                .and_then(|mut file| grammar.export_dot(&name, &mut file))
+                .and_then(|mut file| grammar.export_dot(name, &mut file))
                 .unwrap()
         })
     }
 
     /// Dump the grammar of this database.
-    pub fn print_grammar(&mut self) {
-        time(&self.options.clone(), "print_grammar", || {
-            let name = self.name_result().clone();
-            let grammar = self.grammar_result().clone();
-            grammar.dump(&name);
+    pub fn print_grammar(&self) {
+        time(&self.options, "print_grammar", || {
+            let name = self.name_result();
+            let grammar = self.grammar_result();
+            grammar.dump(name);
         })
     }
 
     /// Dump the formulas of this database.
-    pub fn print_formula(&mut self) {
-        time(&self.options.clone(), "print_formulas", || {
-            let parse = self.parse_result().clone();
-            let name = self.name_result().clone();
-            let stmt_parse = self.stmt_parse_result().clone();
-            stmt_parse.dump(&parse, &name);
+    pub fn print_formula(&self) {
+        time(&self.options, "print_formulas", || {
+            self.stmt_parse_result().dump(self);
         })
     }
 
     /// Verify that printing the formulas of this database gives back the original formulas
-    pub fn verify_parse_stmt(&mut self) {
-        time(&self.options.clone(), "verify_parse_stmt", || {
-            let parse = self.parse_result().clone();
-            let name = self.name_result().clone();
-            let stmt_parse = self.stmt_parse_result().clone();
-            if let Err(diag) = stmt_parse.verify(&parse, &name) {
+    pub fn verify_parse_stmt(&self) {
```

Comment:
For uniformity with the other passes, this shall perform the necessary `pass`.

Reply:
Actually the annotations shall be stored in a separated array and returned like for the other passes. There shall be a `VerifyParseStmtResult`. We can make it a TODO for me.

Reply:
Created issue #40 to track this.